src/query/atomWithQuery.ts

```diff
@@ -60,7 +61,10 @@ export function atomWithQuery<
             }
             action.initializer(getQueryClient(get, set))
           } else if (action.type === 'data') {
-            set(dataAtom, action.data)
+            const data = get(dataAtom)
+            if (!data || !equalityFn(data, action.data)) {
+              set(dataAtom, action.data)
+            }
             const pending = get(pendingAtom)
             if (!pending.fulfilled) {
               pending.resolve(action.data)
```

Comment:
It's probably better to keep the pair of dataAtom and pendingAtom in sync.
```suggestion
            const data = get(dataAtom)
            if (!data || !equalityFn(data, action.data)) {
              set(dataAtom, action.data)
              const pending = get(pendingAtom)
              if (!pending.fulfilled) {
                pending.resolve(action.data)
              }
```